### PR TITLE
support gauges v2

### PR DESCRIPTION
**Pull request by mendesfabio**
Adds support to V2 factories

- https://thegraph.com/hosted-service/subgraph/mendesfabio/balancer-gauges
- https://thegraph.com/hosted-service/subgraph/mendesfabio/balancer-gauges-goerli
- https://thegraph.com/hosted-service/subgraph/mendesfabio/balancer-gauges-arbitrum

### CODE REVIEW

File: schema.graphql [hunk 1]
```diff
@@ -42,59 +88,119 @@ enum Chain {
...
+
+  " Type of the gauge "
   type: GaugeType!
+
+  " Reference to LiquidityGauge "
   liquidityGauge: LiquidityGauge
 }
 
 type Pool @entity {
+  " Pool contract address "
```
Comment (markusbkoch) [inline schema.graphql:122]: " Address of the pool (lp_token of the gauge) "

File: schema.graphql [hunk 2]
```diff
@@ -42,59 +88,119 @@ enum Chain {
...
   id: ID!
+
+  " Pool ID on Balancer's Vault - nullable since it can be any contract "
   poolId: Bytes
+
+  " Pool contract address "
```
Comment (markusbkoch) [inline schema.graphql:128]: " Address of the pool (lp_token of the gauge) "

File: schema.graphql [hunk 3]
```diff
@@ -42,59 +88,119 @@ enum Chain {
...
+
+  " List of gauges created for the pool "
   gauges: [LiquidityGauge!] @derivedFrom(field: "pool")
 }
 
 type RewardToken @entity {
+  " Equal to: <tokenAddress>-<gaugeAddress> "
   id: ID!
+
+  " ERC20 token symbol - empty string if call reverts "
```
Comment (markusbkoch) [inline schema.graphql:142]: " ERC20 token symbol - empty string if call to symbol() reverts "

File: schema.graphql [hunk 4]
```diff
@@ -42,59 +88,119 @@ enum Chain {
...
   symbol: String!
+
+  " ERC20 token decimals - zero if call reverts "
```
Comment (markusbkoch) [inline schema.graphql:145]: " ERC20 token decimals - zero if callt to decimals() reverts "

File: schema.graphql [hunk 5]
```diff
@@ -42,59 +88,119 @@ enum Chain {
...
+  " Equal to: <userAddress>-<gaugeAddress> "
   id: ID!
+
+  " Reference to User entity "
   user: User!
+
+  " Reference to LiquidityGauge entity "
   gauge: LiquidityGauge!
+
+  " Balance of BPTs user has staked in the LiquidityGauge "
```
Comment (markusbkoch) [inline schema.graphql:165]: While it's true that for the existing gauges this is equal to the amount of BPT the user has staked, in reality this is 
  " The user's balance of gauge deposit tokens "

File: schema.graphql [hunk 6]
```diff
@@ -42,59 +88,119 @@ enum Chain {
...
 
 type GaugeVote @entity {
+  " Equal to: <userAddress>-<gaugeAddress> "
   id: ID!
+
+  " Reference to User entity "
   user: User!
+
+  " Reference to LiquidityGauge entity "
   gauge: LiquidityGauge!
```
Comment (markusbkoch) [inline schema.graphql:185]: Shouldn't this be a Gauge entity instead?

**Review by markusbkoch — CHANGES_REQUESTED** [review]
Looks good, requested a few changes in the comments

File: src/gauge.ts [hunk 7]
```diff
@@ -76,14 +76,56 @@ export function handleRootUnkillGauge(call: UnkillGaugeCall): void {
...
+
+  // Update Preferential Gauge
+
+  let poolId = killedGauge.pool;
+  let pool = Pool.load(poolId);
+  if (pool == null) return;
+
+  let preferentialGaugeId = pool.preferentialGauge;
+  if (preferentialGaugeId === null) return;
+  let preferentialGauge = LiquidityGauge.load(preferentialGaugeId) as LiquidityGauge;
```
[the remark this answers is not in the document]
Comment (mendesfabio) [reply src/gauge.ts]: yeah, I changed this a bit and tested the changes tonight [here](https://thegraph.com/hosted-service/subgraph/mendesfabio/balancer-gauges-goerli) - although Subgraph's name says Goerli, it's Mainnet, my bad. Everything is working, see the diff at 316cd214266283f1129d0eb0b03e27e894857f50

File: schema.graphql (hunk 6 above)
Comment (mendesfabio) [reply schema.graphql]: a99fe0f38dc0aaffbf7c6e848c6543f0e452c8ad

File: schema.graphql [hunk 8]
```diff
@@ -42,59 +88,119 @@ enum Chain {
...
+  " Equal to: <gaugeAddress>-<typeID> "
   id: ID!
+
+  " Address of the gauge "
   address: Bytes!
+
+  " Type of the gauge "
   type: GaugeType!
+
+  " Reference to LiquidityGauge "
```
[the remark this answers is not in the document]
Comment (mendesfabio) [reply schema.graphql:117]: f7f12d7f6c1841502b36f2821c80b7ac47f87b89

**Review by markusbkoch — APPROVED** [review]

File: schema.graphql [hunk 9]
```diff
@@ -1,33 +1,83 @@
...
   poolId: Bytes
+
+  " Whether Balancer DAO killed the gauge "
   isKilled: Boolean!
-  streamer: Bytes # ChildChainLiquidityGauge Only
+
+  " Whether Balancer DAO added the gauge to GaugeController "
+  isAdded: Boolean!
+
+  " Timestamp at which Balancer DAO added the gauge to GaugeController [seconds] "
```
Comment (markusbkoch) [inline schema.graphql:62]: " Timestamp at which Balancer DAO added the gauge to GaugeController [seconds] - only applies to mainnet gauges "

File: src/utils/constants.ts [hunk 10]
```diff
@@ -31,3 +31,29 @@ export const CONTROLLER_ADDRESS = forNetwork(controllerAddressByNetwork, network
 export const ARBITRUM_ROOT_GAUGE_FACTORY = Address.fromString('0xad901309d9e9DbC5Df19c84f729f429F0189a633');
 export const OPTIMISM_ROOT_GAUGE_FACTORY = Address.fromString('0x3083A1C455ff38d39e58Dbac5040f465cF73C5c8');
 export const POLYGON_ROOT_GAUGE_FACTORY = Address.fromString('0x4C4287b07d293E361281bCeEe8715c8CDeB64E34');
+
+export const MAINNET_GAUGE_V2_FACTORY = Address.fromString('0x3b8cA519122CdD8efb272b0D3085453404B25bD0');
```
Comment (markusbkoch) [inline src/utils/constants.ts:35]: This is Goerli, need to also add Mainnet as a const and to `isV2Factory`

Comment (markusbkoch) [reply src/utils/constants.ts:35]: Mainnet `0xf1665E19bc105BE4EDD3739F88315cC699cc5b65`

File: schema.graphql [hunk 11]
```diff
@@ -38,49 +88,122 @@ enum Chain {
 }
 
 type RootGauge @entity {
```
Comment (markusbkoch) [inline schema.graphql:90]: Let's also introduce `addedTimestamp` and `factory` attributes in the RootGauge entity

**Review by markusbkoch — CHANGES_REQUESTED** [review]

**Review by markusbkoch — APPROVED** [review]

**Review by markusbkoch — APPROVED** [review]

**Pull request merged**